bagua/torch_api/algorithms/async_model_average.py

```diff
@@ -45,75 +43,183 @@ def __init__(
             peer_selection_mode (str): The way how workers communicate with each other. Currently ``"all"`` is supported.
                 ``"all"`` means all workers' weights are synchronized during each communication.
             sync_interval_ms (int): Number of milliseconds between model synchronizations.
+            warmup_steps (int): Number of steps to warm up by doing gradient allreduce before doing asynchronous
+                model averaging. Use 0 to disable.
         """
 
         self.peer_selection_mode = peer_selection_mode
         self.sync_interval_ms = sync_interval_ms
-        self.stop_event = threading.Event()
-        check_nccl_proto()
+        self.step_id = 0
+        self.warmup_steps = warmup_steps
+
+        self.cuda_event = torch.cuda.Event()
+
+        self.abort_event = threading.Event()
+        self.dummy_tensor = torch.Tensor([0]).byte()
+        self.main_group = torch.distributed.new_group(backend="gloo")
+        self.thread_group = torch.distributed.new_group(backend="gloo")
+
+        self.executor = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+        self.scheduled = False
+
+        if self.warmup_steps <= 0:
+            self.no_bucketing = True
```

Comment:
This seems to be a new API. Why do we need this?

Reply:
Use this to disable bucketing for async algorithms.

It can reduce communication cost compared to the case when we use bucketing, which is verified.

And I think it may improve the performance as well, which it not verified.